### DISCUSSION TITLE
Multiple instances of PrismaClient

### DISCUSSION THREAD

**Discussion by reharik** [prisma/prisma#9718] [Q&A] [answered]
Hi, I am using apollo-server-koa and I am doing authentication in a koa middleware.  I need to instantiate prisma to query the db. Then depending on the route I end up in graphql land.  I have not been able to tell how I can pass the instance of PrismaClient along to graphql so I need to create a new instance of PrismaClient.  Thus in one request I have called new PrismaClient 2 times.  Further I use some prisma middleware so I'm adding that to prisma twice.  
I guess my question is, a) is it ok to instantiate it twice, b) if I do instantiate it twice I presume it's not a singleton so I guess I have to add the middleware twice.  c)  any other ideas?
Thanks,
R

**Comment by ryands17** [comment, marked as answer]
@reharik 👋 

The best practice is to instantiate `PrismaClient` only once, because it creates a connection pool and handles connections in that manner. 

If you're instantiating twice, then you would need to add the middleware twice because the middleware is bound to an instance. So I would recommend only a single instance.

**Reply by reharik** [reply]
This is very helpful.  A follow up question, and maybe I should post separately but, my middleware is contingent on the current user.  Meaning every request needs to "add" the middleware with the current user.  This seems like it might be a problem.  Any thoughts on this?
Thanks

**Reply by reharik** [reply]
ugg.  Seems that there is a middleware stack with no pop so repeatedly adding middleware will just build up a huge stack.  Perhaps I can create a middleware with a callback embeded that changes the current user, then in my auth middleware I can pass the authenticated user to the middleware callback.  I'll have to experiment.
R

**Reply by reharik** [reply]
nope that doesn't make any sense.

**Reply by janpio** [reply]
I suggest you open a new discussion or issue about this - the problem sounds interesting indeed.

Update: Ah, just found https://github.com/prisma/prisma/discussions/9750

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
